### ISSUE TITLE
Force log axes to remain positive

### ISSUE THREAD

**Issue by StevenCHowell** [bokeh/bokeh#5550]
I have noticed when linking the axes of a log plot to a linear plot of the same data, the resulting figure fails gloriously when dragging the linear axis negative.  This is related to issue #5389, #5549, and PR #5477.

I see two ways to prevent problems when users inadvertently designate a situation that does not really make sense for a log plot,
1. impose an axis-limit-override for log scale so the lower limit is ALWAYS greater than zero, i.e., ignore commands to set the limit to zero or negative
2. somehow determine minimum value to show, below the the lowest data value, then mirror the axis to negative, then show that space empty (log(n) > 0 so it would never have data), i.e., the smallest data value is `1e-9` so draw the axis at `1e-10` and mirror across the axis

I am not sure which would be preferred/easier, and there are likely other options worth considering, ideas?

Here is an simple example to demonstrate the strange behavior (note that this was run before PR #5477):
```
import numpy as np
from bokeh.models import Range1d
from bokeh.layouts import gridplot
from bokeh.plotting import figure, show, output_notebook
output_notebook()

x = np.linspace(0.001, 0.9, 300)
y = (np.sin(15 * x) / (15 * x)) ** 2 # fake small-angle scattering data

p1 = figure(title='Linear Scale', x_axis_label='q (1/A)', y_axis_label='I(q)',
            width=400, height=400)
p2 = figure(title='Log Scale', x_axis_label='q (1/A)', y_axis_label='I(q)', 
            width=400, height=400, x_axis_type='log', y_axis_type='log')

p1.circle(x, y)
p2.circle(x, y)

# pre PR #5477 bit to force the log range to be correct
x_range = np.min(x[x > 0]) * 0.9, np.max(x) * 1.1
y_range = np.min(y[y > 0]) * 0.9, np.max(y) * 1.1
p2.x_range = Range1d(*x_range)
p2.y_range = Range1d(*y_range)

# set the ranges to match
p1.x_range = p2.x_range
p1.y_range = p2.y_range

fig = gridplot([[p1, p2]])
show(fig)
```
When dragging the plots, the axis can reverse negative-to-positive, the zoom changes to where you can lose track of where the data is, and possibly other strange behaviors.  Here is the original plot
![image](https://cloud.githubusercontent.com/assets/9484229/21158612/dd23cc06-c14b-11e6-8996-1dbee0646f94.png)

and here is an example of the y-axes inverting
![image](https://cloud.githubusercontent.com/assets/9484229/21158570/b52c7036-c14b-11e6-93cd-e370c35db864.png)

**Comment by bryevdv** [comment]
> impose an axis-limit-override for log scale so the lower limit is ALWAYS greater than zero, i.e., ignore commands to set the limit to zero or negative

can this not already be done by setting `bounds` on the range?

**Comment by StevenCHowell** [comment]
> x_range = np.min(x[x > 0]) * 0.9, np.max(x) * 1.1 y_range = np.min(y[y > 0]) * 0.9, np.max(y) * 1.1 p2.x_range = Range1d(*x_range) p2.y_range = Range1d(*y_range)

I think this code is what you are referring to.  It does set the bounds initially, but things break when you start dragging the linear scale.  

Would it make sense to add a line of code that will verify a new range is positive, else the update is ignored?

**Comment by bryevdv** [comment]
@StevenCHowell no I am referring to the `bounds` property of any range this is intended to limit the range extent. e.g.:
```
plot.x_range.bounds = (0, 10) # plot won't ever go outside (0, 10)
```

**Comment by StevenCHowell** [comment]
@bryevdv I see.  Adding the lines 
```
p2.x_range.bounds = x_range
p2.y_range.bounds = y_range
```
does prevent panning beyond the set range, for both the log and linear plot as the `p1.*_range` equals `p2.*_range`.  

Does it make sense to always set the lower limit to 0 when doing a log plot?  Is it possible to only set one of the limits? It would be simple enough to set the upper limit to something excessively large.

**Comment by StevenCHowell** [comment]
After setting the `*_range.bounds`, I still get some axis inversion when panning the data.  I am not certain what leads to this.  Reseting the image works so maybe this bit is not an important issue.

![image](https://cloud.githubusercontent.com/assets/9484229/21266554/b9c88adc-c374-11e6-893e-aefe78d5c432.png)

**Comment by StevenCHowell** [comment]
I just tested panning the plots generated by the code in my original comment.  In the current dev build, 0.12.5dev11, panning works great.  When the dragging the linear plot beyond the limits of the log plot, the log plot pauses until and the linear plot returns to possible limits of the log plot.  

I also tested the current release build, 0.12.4 (0.12.4-py35_0), and the problems persist there.

As long as nothing gets reverted, I think this safe to close as the 0.12.5 update should resolve the issue.
